### PR TITLE
Added Automatic SpecDM Weapon Generation

### PR DESCRIPTION
**Pull request by markusmarkusz**
I think the meaning explains itself.
I haven't tested any of these changes.

### CODE REVIEW

File: lua/weapons/weapon_ghost_ak47.lua [hunk 1]
```diff
@@ -1,3 +1,5 @@
+AddCSLuaFile()
```
Comment (MinIsMin) [inline lua/weapons/weapon_ghost_ak47.lua]: Why adding AddCSLuaFile() to all files instead of using two in the include function?

Comment (markusmarkusz) [inline lua/weapons/weapon_ghost_ak47.lua]: Because this way is dumb.
And the function of this PR would be broken without using AddCSLuaFile() manually.

**Comment by MinIsMin** [comment]
I like this idea, but I can't get it working:
```
[ULib] addons/specdm/lua/sh_spectator_deathmatch.lua:44: attempt to index local 'wep' (a nil value)
  1. GenerateSpecDMWeapons - addons/specdm/lua/sh_spectator_deathmatch.lua:44
   2. fn - addons/specdm/lua/sh_spectator_deathmatch.lua:67
    3. unknown - lua/ulib/shared/hook.lua:110
```
It shows many empty entries.

Also, please use `string.sub(v,1,` instead of `v.sub(1,`

**Comment by markusmarkusz** [comment]
Try it again.

There is not really any difference between `String:sub()` and `string.sub(String)`.
It just makes the code a bit shorter ^^.

**Comment by MinIsMin** [comment]
Now works without errors.
But no weapons are being added.
I'll look into it tomorrow.

**Comment by markusmarkusz** [comment]
Are you sure?
Try the give command. There are no icons beacause I didn't add them because you haven't done this too.

**Comment by MinIsMin** [comment]
I'm only getting the default ghost weapons.
(Give not tried)

But now I can pick up weapons from the map! xD

**Comment by markusmarkusz** [comment]
Which weapons? Ghost weapons?
I have forgotten some things.

**Comment by MinIsMin** [comment]
Ghost weapons are spawned through the map's spawnlist.
These can be picked up from ghosts.
Also, when sound like PrimarySound are set, they aren't called with "BulletGhost".

**Comment by markusmarkusz** [comment]
Like I have said I have forgotten some small chamges.

**Comment by markusmarkusz** [comment]
Please test it again.

**Comment by MinIsMin** [comment]
No errors are thrown, but the generated weapons aren't added to the tables.
I tried using `weptable = weapons.GetList()` after `GenerateSpecDMWeapons` but it's always nil.

**Comment by MinIsMin** [comment]
Partially working:

I put
```
if wep.Kind == WEAPON_HEAVY then
	table.insert(SpecDM.Ghost_weapons.primary, name)
elseif wep.Kind == WEAPON_PISTOL then
	table.insert(SpecDM.Ghost_weapons.secondary, name)
end
```
after `weapons.Register(wep, name)`, otherwise the generated weapons aren't included.
The AutoIncludeWeapons would be:
```
if SpecDM.AutoIncludeWeapons then
	hook.Add("Initialize", "SharedInitialize_Ghost", function()
		table.Empty(SpecDM.Ghost_weapons.primary)
		table.Empty(SpecDM.Ghost_weapons.secondary)
		local weptable = weapons.GetList()
		if SpecDM.AutoGenerateWeapons then
			GenerateSpecDMWeapons(weptable)
		else
			for _, w in pairs(weptable) do
				if w and w.Kind and w.Base == "weapon_ghost" then
					if w.Kind == WEAPON_HEAVY then
						table.insert(SpecDM.Ghost_weapons.primary, w.ClassName)
					elseif w.Kind == WEAPON_PISTOL then
						table.insert(SpecDM.Ghost_weapons.secondary, w.ClassName)
					end
				end
			end
		end
	end)
end
```

Problems I encouter with this:
- Even the ghost weapons are made to ghost weapons (maybe exclude them, or leave their name normal, the icons are currently not working)
- Some logic errors with pickung up ammo (I'll work on the base to prevent this)

**Comment by markusmarkusz** [comment]
Try again.

**Comment by MinIsMin** [comment]
It's working great. I'll commit in v1.8 when I get the icons working.

**Comment by markusmarkusz** [comment]
The icons are easy. :P

**Comment by markusmarkusz** [comment]
I'll give you some hints for the loadout icons:
SWEP.Icon
SpecDM.LoadoutIcons
weaponclass = icon

**Comment by MinIsMin** [comment]
Thanks for these hints.
I see the problem with custom functions (like shoot or zoom), which I can't get to work.
I don't think it will be a widely used feature because adding a ghost weapon is easy.

**Comment by markusmarkusz** [comment]
> I see the problem with custom functions (like shoot or zoom), which I can't get to work.

Zoom should work. Shoot is replaced with the base function.

**Comment by MinIsMin** [comment]
For some weapons, zoom isn't working (like slowness or in the SecondaryAttack bultin zoom).
For weapons like double barrel shotguns their function is completly broken.

**Pull request closed without merging**